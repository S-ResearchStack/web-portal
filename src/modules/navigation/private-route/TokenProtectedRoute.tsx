import React from 'react';
import { useSelector } from 'react-redux';
import { RouteProps } from 'react-router-dom';

import { isTokenExistSelector } from 'src/modules/auth/auth.slice';
import PrivateRoute from "src/modules/navigation/private-route/PrivateRoute";

type Props = RouteProps;

const TokenProtectedRoute: React.FC<Props> = ({ children, ...rest }) => {
  const authenticated = useSelector(isTokenExistSelector);

  return (
    <PrivateRoute
      authenticated={authenticated}
      {...rest}>
      {children}
    </PrivateRoute>);
};

export default TokenProtectedRoute;
