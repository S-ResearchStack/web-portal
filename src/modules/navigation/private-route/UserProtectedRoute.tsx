import React from 'react';
import { RouteProps } from 'react-router-dom';

import {isUserRegistered} from 'src/modules/auth/auth.slice';
import PrivateRoute from "src/modules/navigation/private-route/PrivateRoute";

type Props = RouteProps;

const UserProtectedRoute: React.FC<Props> = ({  children, ...rest}) => {
  const registered = isUserRegistered();

  return (
    <PrivateRoute
      authenticated={registered}
      {...rest}>
      {children}
    </PrivateRoute>);
};

export default UserProtectedRoute;
