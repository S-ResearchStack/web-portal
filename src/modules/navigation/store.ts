import { createSelector } from '@reduxjs/toolkit';
import { matchPath } from 'react-router-dom';
import { createBrowserHistory, createMemoryHistory, Location } from 'history';

import { RootState } from 'src/modules/store/store';
import { isInsideTest } from 'src/common/utils/testing';

export enum Path {
  Root = '/',

  Registration = '/registration',
  SignIn = '/signin',
  AccountCreate = '/account/create',

  CreateStudy = '/create-study',

  OverviewSubject = '/overview/participants/:subjectNumber',
  StudyManagement = '/study-management',
  StudyManagementSubject = '/study-management/participants/:subjectNumber',
  StudyManagementEditSurvey = '/study-management/surveys/:surveyId/edit',
  StudyManagementSurveyResults = '/study-management/surveys/:surveyId/results',
  StudyManagementActivityResults = '/study-management/activities/:activityId/results',

  Overview = '/overview',
  SubjectManagement = '/subject-management',
  StudyData = '/study-data',
  StudySettings = '/study-settings',
  LabVisitManagement = '/lab-visit',

  Dashboard = '/dashboard',
  CreateChart = '/dashboard/:dashboardId/chart/create',
  EditChart = '/dashboard/:dashboardId/chart/:chartId/edit',

  TaskManagement = '/task-management',
  CreateSurvey = '/task-management/survey/create',
  CreateActivity = '/task-management/activity/:activityType/create',
  EditActivity = '/task-management/activity/:activityId/edit',

  EducationalManagement = '/educational-management',
  CreateEducational = '/educational-management/:educationType/create',
  EditEducational = '/educational-management/:educationId/edit',
}

export const makeHistory = () =>
  !isInsideTest
    ? createBrowserHistory({ basename: process.env.PUBLIC_PATH })
    : createMemoryHistory();

export const history = makeHistory();

const sectionPaths = [
  Path.Overview,
  Path.Dashboard,
  Path.TaskManagement,
  Path.SubjectManagement,
  Path.StudyData,
  Path.EducationalManagement,
  Path.LabVisitManagement,
  Path.StudySettings
] as const;

export const sectionPathSelector = createSelector(
  (state: RootState) => state.router,
  (r) => sectionPaths.find((s) => matchPath((r.location as unknown as Location).pathname, s))
);
