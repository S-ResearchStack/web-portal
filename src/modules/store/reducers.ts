import { combineReducers } from '@reduxjs/toolkit';
import { connectRouter } from 'connected-react-router';
import type { Reducer } from 'redux';

import authReducer from 'src/modules/auth/auth.slice';
import studySettingsReducer from 'src/modules/study-settings/studySettings.slice';
import snackbarReducer from 'src/modules/snackbar/snackbar.slice';
import dashboardReducers from 'src/modules/dashboard/dashboard.slice';
import overviewReducers from 'src/modules/overview/overview.slice';
import studiesReducer from 'src/modules/studies/studies.slice';
import studyManagementReducer from 'src/modules/subject/studyManagement.slice';
import educationManagementReducer from 'src/modules/education-management/educationManagement.slice';
import taskManagementReducer from 'src/modules/task-management/taskManagement.slice';
import labVisitsReducers from 'src/modules/lab-visit/labVisit.slice';
import sidebarReducer from 'src/modules/main-layout/sidebar/sidebar.slice';
import fileUploadReducer from 'src/modules/file-upload/fileUpload.slice';
import { history } from 'src/modules/navigation/store';
import { DataSliceState } from 'src/modules/store/createDataSlice';
import studyDataReducer from 'src/modules/study-data/studyData.slice';
import studyRequirementReducer from "src/modules/studies/ParticipationRequirement.slice"
import participantTimezonesReducers from '../common/participantTimezones.slice';

const createReducer = <T, V>(asyncReducers?: { [key: string]: Reducer<DataSliceState<T, V>> }) =>
  combineReducers({
    ...authReducer,
    router: connectRouter(history),
    snackbar: snackbarReducer,
    sidebar: sidebarReducer,
    ...studySettingsReducer,
    ...dashboardReducers,
    ...overviewReducers,
    ...studyManagementReducer,
    ...educationManagementReducer,
    ...taskManagementReducer,
    ...labVisitsReducers,
    ...fileUploadReducer,
    ...asyncReducers,
    ...studiesReducer,
    ...studyDataReducer,
    ...studyRequirementReducer,
    ...participantTimezonesReducers
  });

export default createReducer;
