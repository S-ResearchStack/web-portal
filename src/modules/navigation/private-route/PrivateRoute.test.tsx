import '@testing-library/jest-dom';
import { act, render, screen } from '@testing-library/react';
import { ConnectedRouter } from 'connected-react-router';
import 'jest-styled-components';
import React from 'react';
import { Provider } from 'react-redux';
import { Route, Switch } from 'react-router-dom';
import { history, Path } from 'src/modules/navigation/store';
import { createTestStore } from '../../store/testing';
import TokenProtectedRoute from './TokenProtectedRoute';
import UserProtectedRoute from './UserProtectedRoute';

describe('PrivateRoute', () => {
  it('should handle authorized', async () => {
    const store = createTestStore({
      auth: {
        authToken: 'token',
      },
    });

    const { baseElement } = await render(
      <Provider store={store}>
        <ConnectedRouter history={history}>
          <UserProtectedRoute path={Path.CreateStudy}>Children</UserProtectedRoute>
        </ConnectedRouter>
      </Provider>
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('[NEGATIVE] should handle NOT authorized', async () => {
    const store = createTestStore({
      auth: {
        authToken: '',
      },
    });

    await act(async () => {
      render(
        <Provider store={store}>
          <ConnectedRouter history={history}>
            <Switch>
              <Route path={Path.Root}>Sign in</Route>
              <UserProtectedRoute>Children</UserProtectedRoute>
            </Switch>
          </ConnectedRouter>
        </Provider>
      );
    });

    expect(screen.queryByText('Children')).toBeNull();
    expect(await screen.findByText('Sign in')).toBeInTheDocument();
  });

  it('[NEGATIVE] should handle empty store', async () => {
    const store = createTestStore({});

    await act(() => {
      render(
        <Provider store={store}>
          <ConnectedRouter history={history}>
            <Switch>
              <Route path={Path.Root}>Sign in</Route>
              <UserProtectedRoute>Children</UserProtectedRoute>
            </Switch>
          </ConnectedRouter>
        </Provider>
      );
    });

    expect(screen.queryByText('Children')).toBeNull();
    expect(await screen.findByText('Sign in')).toBeInTheDocument();
  });
});

describe('TokenProtectedRoute', () => {
  it('should handle authorized', async () => {
    const store = createTestStore({
      auth: {
        authToken: 'token',
      },
    });

    await act(() => {
      render(
        <Provider store={store}>
          <ConnectedRouter history={history}>
            <TokenProtectedRoute>Children</TokenProtectedRoute>
          </ConnectedRouter>
        </Provider>
      );
    });

    expect(await screen.findByText('Children')).toBeInTheDocument();
  });
});
