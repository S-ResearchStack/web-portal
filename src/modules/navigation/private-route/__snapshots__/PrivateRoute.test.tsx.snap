// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateRoute should handle authorized 1`] = `
<body>
  <div />
</body>
`;
